src/rgw/rgw_kafka.cc

```diff
@@ -441,18 +443,16 @@ class Manager {
     // coverity[leaked_storage:SUPPRESS]
   }
 
-  // the managers thread:
-  // (1) empty the queue of messages to be published
-  // (2) loop over all connections and read acks
-  // (3) manages deleted connections
-  // (4) TODO reconnect on connection errors
-  // (5) TODO cleanup timedout callbacks
   void run() noexcept {
     while (!stopped) {
 
       // publish all messages in the queue
       auto reply_count = 0U;
-      const auto send_count = messages.consume_all(std::bind(&Manager::publish_internal, this, std::placeholders::_1));
+      std::vector<message_wrapper_t*> local_messages;
+      const auto send_count = messages.consume_all([&local_messages](auto message) {local_messages.push_back(message);});
```

Comment:
`consume_all` is a non-blocking function and prior to your changes, it was invoking `publish_internal` which was run async in diff thread where the `conn` was used. and while below code had a block where it checks to `conn` to be idle and if yes it deletes the connection.
so there were chances of race condition where before `publish_internal` goes and updates the `conn->time` the `idle` check was executed and conn was destroyed causing the race condition and crash when `publish_internal` tries to access the `conn`
if yes, i just have 1 doubt, `consume_all` will still run async and `local_messages` will not be populated immediately and then `std::for_each` would loop over the `local_messages` which would not have populated and this will result in loss of messages being delivered?
with current changes, you are consuming all the messages and then calling the `publish_internal` in `std::for_each` loop so ensuring the `idle` check will be executed only after the `publish_internal` as been completed for all of the messages and then this would avoid the race condition as destroy and publish_internal would never be executed in parallel?

Reply:
we were running in the same thread (the kafka manager thread) even before the change.
nothing was running in paralel, and `consume_all` is completly syncronouse, so, i don't think there was a risk for race condition.
the main reason for the change is that before the change, traversing the queue was taking longer (since we called `publish_internal` as we traversed the queue.
this means that it is more likely for the thread that pushes into the queue to find it full.
however, i did not see any visible performence improvments with the change, so i'm also ok with reverting it.

Reply:
probably reverting the change would make sense then for `consume_all` ??

Reply:
done